plugins/lookup/etcd3.py

```diff
@@ -0,0 +1,185 @@
+        ca_cert:
+            description:
+            - etcd3 CA authority
+            default: None
+            env:
+            - name: ETCDCTL_CACERT
+            type: string
+        cert_cert:
+            description:
+            - etcd3 client certificate
+            default: None
+            env:
+            - name: ETCDCTL_CERT
+            type: string
+        cert_key:
+            description:
+            - etcd3 client private key
+            default: None
+            env:
+            - name: ETCDCTL_KEY
+            type: string
+        timeout:
+            description:
+            - client timeout
+            default: 60
+            type: integer
+        user:
+            description:
+            - auth user name
+            default: None
+            type: string
+        password:
+            description:
+            - auth user password
+            default: None
+            type: string
+
+    requirements:
+    - "etcd3 >= 0.12"
+'''
+
+EXAMPLES = '''
+    - name: "a value from a locally running etcd"
+      debug: msg={{ lookup('etcd3', 'foo/bar') }}
+
+    - name: "values from multiple folders on a locally running etcd"
+      debug: msg={{ lookup('etcd3', 'foo', 'bar', 'baz') }}
+
+    - name: "look for a key prefix"
+      debug: msg="{{ lookup('etcd3', '/foo/bar', prefix=True) }}"
+
+    - name: "connect to etcd3 with a client certificate"
+      debug msg: "{{ lookup('etcd3', 'foo/bar', cert_cert='/etc/ssl/etcd/client.pem', cert_key='/etc/ssl/etcd/client.key') }}"
+'''
+
+RETURN = '''
+    _raw:
+        description:
+        - list of keys and associated values
+        type: list
+        elements: complex
+'''
+
+import os
+import traceback
+
+from ansible.plugins.lookup import LookupBase
+from ansible.module_utils.basic import missing_required_lib
+from ansible.module_utils._text import to_native
+
+try:
+    import etcd3
+    HAS_ETCD = True
+except ImportError:
+    ETCD_IMP_ERR = traceback.format_exc()
+    HAS_ETCD = False
+
+
+class LookupModule(LookupBase):
+
+    def run(self, terms, variables, **kwargs):
+
+        self.set_options(var_options=variables, direct=kwargs)
+
+        if not HAS_ETCD:
+            self._display.error(missing_required_lib('etcd3'))
+            return None
+
+        # get etcd3 optional args
+        client_params = {
+            'host': '127.0.0.1',
+            'port': 2379,
+            'ca_cert': None,
+            'cert_cert': None,
+            'cert_key': None,
+            'timeout': 60,
+            'user': None,
+            'password': None
+        }
+
+        # use ETCD environment variables if they are set
+        etcd3_envs = {
+            'ca_cert': 'ETCDCTL_CACERT',
+            'cert_cert': 'ETCDCTL_CERT',
+            'cert_key': 'ETCDCTL_KEY',
+        }
+        for key in etcd3_envs.keys():
+            value = os.getenv(etcd3_envs[key])
+            if value:
+                client_params[key] = value
```

Comment:
Please remove this code block (lines 144-153). Ansible handles the env variables as documented in the module options.

Reply:
fixed in commit e57a7f6d32c1738a6e89dabb4b500016def88c23